File: 오주영/Week2/BOJ_10828_스택.py
import sys  #입력을 stdin 형태로 받기 위해 불러온다.


S = []

#Steak 기능 구현 ~ list를 이용해서
def push(S : list, any):    #레퍼런스에서는 any 대신 f, f[1]을 사용했다. 장점이 있나?
    d = int(any)
    S.append(d)             #데이터를 가장 오른쪽에 넣는다.

def pop(S):
    if S == []:
        print(-1)
    else:
        value = S[len(S)-1] #S의 가장 오른쪽 값이다 /레퍼런스 때문에 직접 print를 씌우지 않았는데 이유가 궁금하다(아마도 활용을 위해?)
        S.pop()             #S에서 가장 오른쪽 값을 지운다
        print(value)

def top(S):
    if S == []:
        print(-1)
    else:
        print(S[len(S)-1])

def empty(S):
    if S==[]:
        print(1)
    else:
        print(0)

def size(S):
    print(len(S))




N = int(sys.stdin.readline()) #문자열을 int로 형변환 하면 개행문자 지워짐

for i in range(N):
    f = sys.stdin.readline().split() #문자열을 공백단위로 / f[0]이 명령어가 된다

    if f[0]=='push':
        push(S,f[1])
    elif f[0]=='pop':
        pop(S)
    elif f[0]=='top':
        top(S)
    elif f[0]=='empty':
        empty(S)
    elif f[0]=='size':
        size(S)
    
    '''함수명과 f[0]가 같은데 코드를 더 짧게 하는 방법이 있을까?'''

File: 성경민/Week2/BOJ_10828_스택.java
10828(스택)

push X: 정수 X를 스택에 넣는 연산이다.
pop: 스택에서 가장 위에 있는 정수를 빼고, 그 수를 출력한다. 만약 스택에 들어있는 정수가 없는 경우에는 -1을 출력한다.
size: 스택에 들어있는 정수의 개수를 출력한다.
empty: 스택이 비어있으면 1, 아니면 0을 출력한다.
top: 스택의 가장 위에 있는 정수를 출력한다. 만약 스택에 들어있는 정수가 없는 경우에는 -1을 출력한다.

pop
top
push 123
top
pop
top
pop


import java.io.*;
import java.util.*;

public class Main{
    public static void main(String[] args)throws NumberFormatException, IOException {
		
		// stack 객체 선언
        Stack<Integer> s = new Stack<>();

		//BufferedReader, BufferedWriter 객체 선언
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		
		// 첫째 줄에 부여하는 명령의 수 n으로 지정
		// scanner 와 다르게 readLine은 String 형태 이므로 형 변환을 해줘야 함
        int x = Integer.parseInt(br.readLine());
		
		// x만큼 반복
        for (int i = 0; i < x; i++) {
			// 명령은 push 123 형태, .split(" ")을 사용 함
			String[] str = br.readLine().split(" ");
			
			// 스위치 문으로 명령마다 행위을 정의해 준다.
            switch (str[0]) {
                case "push":
					
					// 정수 str[1]을 스택에 넣는 연산
                    s.push(Integer.parseInt(str[1]));
                    break;

                case "pop":
				
					// 스택에서 가장 위에 있는 정수를 빼고 그 수를 출력한다.
					// 만약 스택이 비어있을 경우 -1을 출력
                    if (s.empty()) {
					// write함수는 개행문자를 포함하지 안아 줄바꿈을 해주거나 bw.newLine();을 추가해서 줄바꿈 진행 해야한다.
                        bw.write("-1" + "\n");
                    } else {
                        bw.write(s.pop() + "\n");
                    }
                    break;

                case "size":
					
					// 스택에 들어있는 정수 개수 출력
                    bw.write(s.size() + "\n");
                    break;

                case "empty":
				
					// 비어있으면 1출력
                    if (s.empty()) {
                        bw.write("1" + "\n");
                    } else {
                        bw.write("0" + "\n");
                    }
                    break;

                case "top":
				
					// 스택에 가장 위에있는 정수를 출력, 비어있으면 -1 출력
                    if(s.empty()){
                        bw.write("-1"+"\n");
                    } else {
					// 꺼내는게 아닌 출력만 하므로 peek 
                        bw.write(s.peek()+"\n");
                    }
                    break;
            }
        }
        bw.flush();	// 버퍼에 남아 있는 문자열 출력해서 버퍼를 지움
        bw.close(); // 스트림 닫기
    }
}
